records_mover/url/gcs/gcs_directory_url.py

```diff
@@ -0,0 +1,85 @@
+from urllib.parse import urlparse, unquote
+from records_mover.url import BaseDirectoryUrl, BaseFileUrl
+import google.auth.credentials
+import google.cloud.storage
+import googleapiclient.discovery
+from typing import List, TYPE_CHECKING
+if TYPE_CHECKING:
+    from .gcs_file_url import GCSFileUrl
+
+
+class GCSDirectoryUrl(BaseDirectoryUrl):
+    def __init__(self,
+                 url: str,
+                 gcs_client: google.cloud.storage.Client,
+                 gcp_credentials: google.auth.credentials.Credentials,
+                 **kwargs) -> None:
+        self.url = url
+        parsed = urlparse(url)
+        self.blob = unquote(parsed.path[1:])
+        self.scheme = parsed.scheme
+        self.bucket = parsed.netloc
+        self.client = gcs_client
+        self.credentials = gcp_credentials
+        self.bucket_obj = self.client.bucket(self.bucket)
+
+    def _directory(self, url: str) -> 'GCSDirectoryUrl':
+        return GCSDirectoryUrl(url, gcs_client=self.client, gcp_credentials=self.credentials)
+
+    def _file(self, url: str) -> 'GCSFileUrl':
+        from .gcs_file_url import GCSFileUrl
+        return GCSFileUrl(url, gcs_client=self.client, gcp_credentials=self.credentials)
+
+    def directory_in_this_directory(self, directory_name: str) -> 'GCSDirectoryUrl':
+        return self._directory(f"{self.url}{directory_name}/")
+
+    def purge_directory(self) -> None:
+        for dir_loc in self.directories_in_directory():
+            dir_loc.purge_directory()
+        for file_loc in self.files_in_directory():
+            file_loc.delete()
+        self.bucket_obj.blob(self.blob).delete()
+
+    def files_in_directory(self) -> List[BaseFileUrl]:
+        prefix = self.blob
+        if prefix == '/':
+            # TODO: Does this ever come out as '/'?  If so, doe sit need to be stripped from others?
+            # API doesn't seem to recognize the root prefix as anything other than ''
+            prefix = ''
+
+        blobs = self.client.list_blobs(bucket_or_name=self.bucket,
+                                       prefix=prefix, delimiter='/')
+        blob_names = [
+            blob.name
+            for blob in blobs
+            # I've seen this happen with gs://bluelabs-test-recordsmover/bar/
+            if blob.name != prefix
+        ]
+        file_urls = [
+            f"gs://{self.bucket}/{blob_name}"
+            for blob_name in blob_names
+        ]
+        return [self._file(file_url) for file_url in file_urls]
+
+    def directories_in_directory(self) -> List[BaseDirectoryUrl]:
+        # https://stackoverflow.com/a/57099354/9795956
+
+        service = googleapiclient.discovery.build('storage', 'v1',
+                                                  credentials=self.credentials)
+        prefix = self.blob
+        if prefix == '/':
+            # TODO: Does this ever come out as '/'?  If so, doe sit need to be stripped from others?
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker